# Committee-Candidate Association &ndash; Data Quality #

## Overview ##

The `dq1` and `dq2` notebooks in this same directory examined the "quality" of the Committee and Candidate FEC data sets, respectively.  This current notebook takes a look at the nature and integrity of the association between the two.  There are actually three different relational structures that can represent connections between Committees and Candidates:

* Many-to-one relationship, through the `cmte.cand_id` foreign key
    * Note that this key only pertains to `cmte_tp` value of "H", "S", or "P"
* One-to-many relationship, through the `cand.cand_pcc` foreign key
    * Note: "pcc" stands for "principal campaign committee"
* Many-to-many relationship, through the `cand_cmte` intersect ("link") table
* Many-to-one relationship, through the `cmte_contrib` table (which itself is a many-to-many between `cmte` and itself)

For now, we will only consider the first three of these association mechanism.  We'll take a look at the fourth relationship if/when we start drilling down on Committee Contributions (i.e. the data in `cmte_contrib`).

Thus, here is a list of the examinations in this notebook:

* Integrity of `cmte.cand_id` foreign key
* Integrity of `cand.cand_pcc` foreign key
* Integrity of `cand_cmte` intersect table
* Consistency of Committees to Candidates, through `cmte.cand_id` foreign key
* Cardinality of Committees to Candidates, through `cand_cmte` intersect table
* Cardinality of Candidates to Committees, through `cmte.cand_id` foreign key
* Cardinality of Candidates to Committees, through `cand_cmte` intersect table

## Notebook Setup ##

### Configure database connect info/options ###

Note: database connect string can be specified on the initial `%sql` command:

```python
database_url = "postgresql+psycopg2://user@localhost/fecdb"
%sql $database_url

```

Or, connect string is taken from DATABASE_URL environment variable (if not specified for `%sql`):

```python
%sql

```

In [1]:
%load_ext sql
%config SqlMagic.autopandas=True
%config InteractiveShell.ast_node_interactivity='last_expr_or_assign'
# connect string taken from DATABASE_URL environment variable
%sql

'Connected: crash@fecdb'

### Configure Python modules ###

In [2]:
import pandas as pd

pd.set_option("display.max_rows", 200)

### Set styling ###

In [3]:
%%html
<style>
  tr, th, td {
    text-align: left !important;
  }
</style>

## Examination of Committee-Candidate Associations ##

### High-level summary ###

#### Replay `cmte` summary stats ####

In [4]:
%%sql result <<
select count(*) as count_total,
       count(distinct cmte_id) as count_distinct_ids,
       count(distinct cmte_nm) as count_distinct_names
  from cmte

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.
Returning data to local variable result


In [5]:
cmte_count_total    = int(result.loc[0][0])
cmte_distinct_ids   = int(result.loc[0][1])
cmte_distinct_names = int(result.loc[0][2])
"cmte_count_total = %d, cmte_distinct_ids = %d, cmte_distinct_names = %d" % \
    (cmte_count_total, cmte_distinct_ids, cmte_distinct_names)

'cmte_count_total = 138384, cmte_distinct_ids = 45868, cmte_distinct_names = 50614'

#### Replay `cand` summary stats ####

In [6]:
%%sql result <<
select count(*) as count_total,
       count(distinct cand_id) as count_distinct_ids,
       count(distinct cand_name) as count_distinct_names
  from cand

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.
Returning data to local variable result


In [7]:
cand_count_total    = int(result.loc[0][0])
cand_distinct_ids   = int(result.loc[0][1])
cand_distinct_names = int(result.loc[0][2])
"cand_count_total = %d, cand_distinct_ids = %d, cand_distinct_names = %d" % \
    (cand_count_total, cand_distinct_ids, cand_distinct_names)

'cand_count_total = 56615, cand_distinct_ids = 26243, cand_distinct_names = 26807'

#### Gather `cand_cmte` summary stats ####

Record total records and count of distinct foreign key values on the `cand_cmte` intersect table

In [8]:
%%sql result <<
select count(*) as count_total,
       count(distinct cmte_id) as count_cmte,
       count(distinct cand_id) as count_cand
  from cand_cmte

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.
Returning data to local variable result


In [9]:
cand_cmte_count_total = int(result.loc[0][0])
cand_cmte_count_cmte  = int(result.loc[0][1])
cand_cmte_count_cand  = int(result.loc[0][2])
"cand_cmte_count_total = %d, cand_cmte_count_cmte = %d, cand_cmte_count_cand = %d" % \
    (cand_cmte_count_total, cand_cmte_count_cmte, cand_cmte_count_cand)

'cand_cmte_count_total = 55899, cand_cmte_count_cmte = 22305, cand_cmte_count_cand = 18291'

In [10]:
"cmte_ids represented: %.2f%%, cand_ids represented: %.2f%%" % \
    (cand_cmte_count_cmte / cmte_distinct_ids * 100.0,
     cand_cmte_count_cand / cand_distinct_ids * 100.0)

'cmte_ids represented: 48.63%, cand_ids represented: 69.70%'

### Integrity of `cmte.cand_id` foreign key ###

Let's check the number (and percentage) of null `cand_id`'s, which presumably represent Committees with no direct Candidate association (may have one or more associations through the `cand_cmte` intersect table)

In [11]:
%%sql
select count(*) as null_cand_ids,
       round(count(*)::numeric / :cmte_count_total * 100.0, 2) as pct_null_cand_ids
  from cmte
 where cand_id is null

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


null_cand_ids pct_null_cand_ids
0          86964             62.84

Same thing, but only looking at distinct `cmte_id`'s (i.e. across election cycles)

In [12]:
%%sql
select count(distinct cmte_id) as distinct_null_cand_ids,
       round(count(distinct cmte_id)::numeric / :cmte_distinct_ids * 100.0, 2) as pct_distinct_null_cand_ids
  from cmte
 where cand_id is null

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


distinct_null_cand_ids pct_distinct_null_cand_ids
0                   23865                      52.03

Now let's look for `cand_id`'s that don't join to Candidate records from the same election cycle (`cand` and `cmte` files should be published together by election cycle)

In [13]:
%%sql
select count(*)
  from cmte
 where (cand_id, elect_cycle) not in (select cand_id, elect_cycle from cand)

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


count
0     18

Of those, let's see how many of the `cand_id`'s just point into outer space (i.e. don't exist in *any* election cycle's Candidate data)

In [14]:
%%sql
select count(*)
  from cmte
 where (cand_id) not in (select cand_id from cand)

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


count
0     11

Let's take a look at those, and see if there are perhaps other Committee records with the same `cmte_id` that
*do* join to Candidate records

In [15]:
%%sql
select cmte.cmte_id,
       cmte.cmte_nm,
       count(*),
       array_agg(distinct cmte.elect_cycle) as elect_cycles,
       array_agg(distinct cmte.cand_id)     as bad_cand_ids,
       array_agg(distinct cand2.cand_id)    as other_cand_ids
  from cmte
  left join cmte cmte2 on  cmte2.cmte_id     = cmte.cmte_id
  left join cand cand2 on  cand2.cand_id     = cmte2.cand_id
                       and cand2.elect_cycle = cmte2.elect_cycle
 where (cmte.cand_id) not in (select cand_id from cand)
 group by 1, 2
 order by 1, 2

 * postgresql+psycopg2://crash@caladan/fecdb
8 rows affected.


cmte_id                                         cmte_nm  count  \
0  C00356071  BILL HAAS FOR CONGRESS (1ST DISTRICT MISSOURI)      4   
1  C00363937                     MIKE GALLAGHER FOR CONGRESS      2   
2  C00625046                COMMITTEE TO ELECT BENJAMIN NOFS      4   
3  C00626838                     SHELLY SCHRATZ FOR CONGRESS      1   
4  C00650424                       DWIGHT BRADY FOR CONGRESS      1   
5  C00677690               WASHINGTON FOR CONGRESS COMMITTEE      4   
6  C00712851                       AMERICAN RELIGIOUS  PARTY      1   
7  C00716191                        LOGAN WILDE FOR CONGRESS      1   

   elect_cycles bad_cand_ids other_cand_ids  
0  [2000, 2002]  [H0MO01082]         [None]  
1        [2000]  [H0FL15047]         [None]  
2  [2016, 2018]  [H6MD10024]         [None]  
3        [2016]  [H6NY26123]         [None]  
4        [2018]  [H8MI01156]         [None]  
5  [2018, 2020]  [H8PA13158]         [None]  
6        [2020]  [C00713602]         [None]  
7        [2020]  [H0UT01122]         [None]

Now we'll take a look at the Committee-Candidate joins (through `cmte.cand_id`) that don't match within the election cycles they came with, but do join across disparate election cycles

In [16]:
%%sql
select count(*)
  from cmte
 where (cand_id, elect_cycle) not in (select cand_id, elect_cycle from cand)
   and (cand_id) in (select cand_id from cand)

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


count
0      7

Let's take a look at those records

In [17]:
%%sql
select cmte.cmte_id,
       cmte.cmte_nm,
       cmte.elect_cycle as cmte_ec,
       cmte.cand_id,
       array_agg(distinct cand.cand_name) as cand_names,
       array_agg(distinct cand.elect_cycle) as cand_elect_cycles
  from cmte
  join cand on cand.cand_id = cmte.cand_id
 where (cmte.cand_id, cmte.elect_cycle) not in (select cand_id, elect_cycle from cand)
 group by 1, 2, 3, 4
 order by 1, 2, 3, 4

 * postgresql+psycopg2://crash@caladan/fecdb
7 rows affected.


cmte_id                             cmte_nm  cmte_ec    cand_id  \
0  C00330043       ROSKAM FOR CONGRESS COMMITTEE     2000  H6IL06117   
1  C00342261      PEOPLE FOR AMERICAN LEADERSHIP     2000  P00003392   
2  C00390468    DRAFT HILLARY FOR PRESIDENT 2004     2004  P00003392   
3  C00400473  HILLARY CLINTON FOR PRESIDENT 2008     2004  P00003392   
4  C00459768    JOSE RUIZ FOR CONGRESS COMMITTEE     2012  H0FL19049   
5  C00636308       GABRIEL MCARTHUR FOR CONGRESS     2018  H8C006211   
6  C00636308       GABRIEL MCARTHUR FOR CONGRESS     2020  H8C006211   

                                          cand_names  \
0                                    [ROSKAM, PETER]   
1  [CLINTON, HILLARY RODHAM, CLINTON, HILLARY ROD...   
2  [CLINTON, HILLARY RODHAM, CLINTON, HILLARY ROD...   
3  [CLINTON, HILLARY RODHAM, CLINTON, HILLARY ROD...   
4                                     [RUIZ, JOSE M]   
5                      [MCARTHUR, GABRIEL SHAWN MR.]   
6                      [MCARTHUR, GABRIEL SHAWN MR.]   

                                  cand_elect_cycles  
0  [2006, 2008, 2010, 2012, 2014, 2016, 2018, 2020]  
1  [2006, 2008, 2010, 2012, 2014, 2016, 2018, 2020]  
2  [2006, 2008, 2010, 2012, 2014, 2016, 2018, 2020]  
3  [2006, 2008, 2010, 2012, 2014, 2016, 2018, 2020]  
4                          [2010, 2014, 2016, 2018]  
5                                            [2016]  
6                                            [2016]

### Integrity of `cand.cand_pcc` foreign key ###

First, let's take a look at percentage of `cand` records with `cand_pcc` specified, and percentage of those that successfully join to `cmte` records, summarized by `cand_status`

In [18]:
%%sql
select cn.cand_status,
       count(*)                          as count,
       round(count(*)::numeric / :cand_count_total * 100.0, 2)
                                         as pct_of_total,
       count(cn.cand_pcc)                as pcc_cmte_specified,
       round(count(cn.cand_pcc)::numeric / count(*) * 100.0, 2)
                                         as pct_of_count,
       count(cn.cand_pcc) - count(cm.id) as pcc_cmte_not_found,
       round((count(cn.cand_pcc) - count(cm.id))::numeric / count(cn.cand_pcc) * 100.0, 2)
                                         as pct_of_specified
  from cand cn
  left join cmte cm on  cm.cmte_id     = cn.cand_pcc
                    and cm.elect_cycle = cn.elect_cycle
 group by 1
 order by 2 desc

 * postgresql+psycopg2://crash@caladan/fecdb
6 rows affected.


cand_status  count pct_of_total  pcc_cmte_specified pct_of_count  \
0           N  24868        43.92               14717        59.18   
1           C  17504        30.92               17452        99.70   
2           P  13617        24.05               13580        99.73   
3           F    617         1.09                 615        99.68   
4        None      8         0.01                   8       100.00   
5           A      1         0.00                   1       100.00   

   pcc_cmte_not_found pct_of_specified  
0                 595             4.04  
1                  20             0.11  
2                 279             2.05  
3                   1             0.16  
4                   0             0.00  
5                   1           100.00

For the ones that join, let's see how many have a non-reciprocal `cand_id` in the PCC `cmte` table

In [19]:
%%sql
select cn.cand_status,
       count(*),
       sum((cm.cand_id != cn.cand_id)::integer) as mismatches,
       round(sum((cm.cand_id != cn.cand_id)::integer)::numeric / count(*) * 100.0, 2) as pct_mismatch
  from cand cn
  join cmte cm on  cm.cmte_id     = cn.cand_pcc
               and cm.elect_cycle = cn.elect_cycle
 group by 1
 order by 2 desc

 * postgresql+psycopg2://crash@caladan/fecdb
5 rows affected.


cand_status  count  mismatches pct_mismatch
0           C  17432          84         0.48
1           N  14122          91         0.64
2           P  13301         217         1.63
3           F    614           1         0.16
4        None      8           0         0.00

And let's list some of those non-reciprocal Candidate records&mdash;first, those whose PCC `cand_id` joins back to a (different) Candidate record with a matching name (only looking at `cand_status` = "C" for now)

In [20]:
%%sql
select cn.elect_cycle,
       cn.cand_id,
       cn.cand_name,
       cn.cand_pcc,
       cm.cmte_nm,
       cm.cand_id as cmte_cand_id,
       cn2.cand_name as cmte_cand_name
  from cand cn
  join cmte cm       on  cm.cmte_id      = cn.cand_pcc
                     and cm.elect_cycle  = cn.elect_cycle
  left join cand cn2 on  cn2.cand_id     = cm.cand_id
                     and cn2.elect_cycle = cm.elect_cycle
 where cn.cand_status in ('C')
   and cm.cand_id != cn.cand_id
   and cn2.cand_name = cn.cand_name
 order by 3, 1
 limit 100

 * postgresql+psycopg2://crash@caladan/fecdb
52 rows affected.


elect_cycle    cand_id                        cand_name   cand_pcc  \
0          2012  H0MO02148                     AKIN, W TODD  C00343475   
1          2008  H6ME01157                  ALLEN, THOMAS H  C00328245   
2          2010  H6ME01157                  ALLEN, THOMAS H  C00328245   
3          2012  S0NV00138                 ANGLE, SHARRON E  C00460758   
4          2012  H8WI00018                   BALDWIN, TAMMY  C00326801   
5          2018  H2PA11098                    BARLETTA, LOU  C00445122   
6          2020  P00007492  BENZEL, JULIANNE ELIZABETH MRS.  C00676320   
7          2012  H8NV01071                 BERKLEY, SHELLEY  C00325738   
8          2018  H2TN06030           BLACKBURN, MARSHA MRS.  C00376939   
9          2010  H6MO07128                       BLUNT, ROY  C00304758   
10         2006  H2OH13033                   BROWN, SHERROD  C00264697   
11         2014  S0CO00237                  BUCK, KENNETH R  C00461368   
12         2012  S2NY00267                   CARVIN, JOSEPH  C00515601   
13         2010  H2DE00072                CASTLE, MICHAEL N  C00254938   
14         2014  H2AR04083                   COTTON, THOMAS  C00499988   
15         2018  H0ND01026                CRAMER, KEVIN MR.  C00504704   
16         2012  S2MT00096                   DAINES, STEVEN  C00491357   
17         2014  H2MT01060                   DAINES, STEVEN  C00491357   
18         2010  H6IN08210                  ELLSWORTH, BRAD  C00412346   
19         2018  S8MI00356               EPSTEIN, LENA ROSE  C00641498   
20         2010  S0CO00229                  FRAZIER, RYAN L  C00461020   
21         2014  H0CO04122                    GARDNER, CORY  C00492454   
22         2016  S6FL00376               GRAYSON, ALAN MARK  C00424713   
23         2006  H2FL13113                HARRIS, KATHERINE  C00369447   
24         2018  H4WV03070                  JENKINS, EVAN H  C00548271   
25         2010  H0IL10120                KIRK, MARK STEVEN  C00350785   
26         2000  H2NY02057                    LAZIO, RICK A  C00323618   
27         2012  H4FL14059                     MACK, CONNIE  C00391243   
28         2004  H2GA04110                MAJETTE, DENISE L  C00373290   
29         2016  S6IL00185                 MARSHALL, ROBERT  C00583567   
30         2018  H8MD02118                      MATORY, LIZ  C00583781   
31         2000  H0FL05014                   MCCOLLUM, BILL  C00166041   
32         2006  H2NJ13075                 MENENDEZ, ROBERT  C00264564   
33         2018  H0IN02125              MESSER, ALLEN LUCAS  C00460667   
34         2014  H8MI09068                     PETERS, GARY  C00437889   
35         2018  H0OH16097                RENACCI, JAMES B.  C00466359   
36         2012  S2MA00147        ROBINSON, HERBERT WILLIAM  C00451146   
37         2018  H0IN04170          ROKITA, THEODORE EDWARD  C00476192   
38         2018  H4MT00050                  ROSENDALE, MATT  C00548289   
39         2018  H6NV03139                     ROSEN, JACKY  C00606939   
40         2018  H2AZ09019                  SINEMA, KYRSTEN  C00508804   
41         2002  S6NJ00164                  SMITH, ROBERT G  C00303552   
42         2016  S6NY00383                  SPOTORNO, FRANK  C00579599   
43         2016  H0IN03198               STUTZMAN, MARLIN A  C00484683   
44         2010  S8NH00139                   SWETT, KATRINA  C00431502   
45         2012  S0NV00203                 TARKANIAN, DANNY  C00465963   
46         2010  S6NY00268                 TASINI, JONATHAN  C00462598   
47         2008  H8NM03097                       UDALL, TOM  C00329896   
48         2012  S0TX00142              WILLIAMS, MICHAEL L  C00457960   
49         2014  S4IA00103                     YOUNG, DAVID  C00545616   
50         2016  H0IN09070          YOUNG, TODD CHRISTOPHER  C00459255   
51         2014  S4GA11269                  YU, EUGENE CHIN  C00547158   

                                    cmte_nm cmte_cand_id  \
0                    

And now those for which the Candidate specified by the PCC's `cand_id` has a *different* name than the original Candidate record (again, only looking at `cand_status` = "C" for now)

In [21]:
%%sql
select cn.elect_cycle,
       cn.cand_id,
       cn.cand_name,
       cn.cand_pcc,
       cm.cmte_nm,
       cm.cand_id as cmte_cand_id,
       cn2.cand_name as cmte_cand_name
  from cand cn
  join cmte cm       on  cm.cmte_id      = cn.cand_pcc
                     and cm.elect_cycle  = cn.elect_cycle
  left join cand cn2 on  cn2.cand_id     = cm.cand_id
                     and cn2.elect_cycle = cm.elect_cycle
 where cn.cand_status in ('C')
   and cm.cand_id != cn.cand_id
   and cn2.cand_name != cn.cand_name
 order by 3, 1
 limit 100

 * postgresql+psycopg2://crash@caladan/fecdb
31 rows affected.


elect_cycle    cand_id                           cand_name   cand_pcc  \
0          2018  H8MD03116               ADAM DAVIDSON DEMARCO  C00660282   
1          2010  S0CO00302                 BARTON, STEVEN KENT  C00468710   
2          2016  H4LA07029        BOUSTANY, CHARLES W. DR. JR.  C00394866   
3          2014  H4TN04130                           CARR, JOE  C00541904   
4          2018  H2MD06195                     DELANEY, JOHN K  C00508416   
5          2012  H4IN02101          DONNELLY, JOSEPH SIMON MR.  C00393652   
6          2012  H0AZ01184                     FLAKE, JEFF MR.  C00347260   
7          2006  H6TN09043                     FORD, HAROLD JR  C00316141   
8          2010  H6NY20167  GILLIBRAND, KIRSTEN ELIZABETH MRS.  C00413914   
9          2014  H2GA11149                  GINGREY, PHIL REP.  C00370783   
10         2012  H8NM01224                    HEINRICH, MARTIN  C00434563   
11         2012  H6HI02251                  HIRONO, MAZIE MRS.  C00420760   
12         2018  H8NC08109                      HUFFMAN, SCOTT  C00663229   
13         2016  S6FL00350                      JOLLY, DAVID W  C00551572   
14         2010  H6NE03131                        KLEEB, SCOTT  C00414656   
15         2014  H0OK05114            LANKFORD, JAMES PAUL MR.  C00466482   
16         2012  H8CA47052               LAUTEN, ROBERT EDMUND  C00496554   
17         2012  P20004669        LOPEZ, CHRISTINA (VICE PRES)  C00512343   
18         2020  H8NM03196                    LUJAN, BEN R MR.  C00443689   
19         2016  P60006954               LYNCH, DENNIS MICHAEL  C00576074   
20         2018  S2OH00170                        MANDEL, JOSH  C00494930   
21         2010  H4LA03028                 MELANCON, CHARLES J  C00397034   
22         2016  H6IL18153                      MELLON, ROBERT  C00582460   
23         2016  H6FL01143                         MILLS, JOHN  C00565366   
24         2018  H2TX16185               O'ROURKE, ROBERT BETO  C00501197   
25         2006  H2VT00086                PARKE, GREGORY TARLE  C00390195   
26         2014  S4MT00084                  STAPLETON, COREY C  C00542068   
27         2002  H6SD00085                       THUNE, JOHN R  C00307165   
28         2008  H8CO02087                      UDALL, MARK E.  C00331439   
29         2008  H8NM01174                  WILSON, HEATHER A.  C00334060   
30         2018  S8PA00338            WRIGHT, THERESA MICHELLE  C00659953   

                                              cmte_nm cmte_cand_id  \
0                     COMMITTEE TO ELECT ADAM DEMARCO    H8DC03014   
1                           STEVE BARTON CAMPAIGN INC    H0CO01102   
2                             BOUSTANY FOR SENATE INC    S6LA00300   
3                                 JOE CARR FOR SENATE    S4TN00302   
4                             FRIENDS OF JOHN DELANEY    P00006213   
5                                DONNELLY FOR INDIANA    S2IN00091   
6                        JEFF FLAKE FOR US SENATE INC    S2AZ00141   
7                        HAROLD FORD JR FOR TENNESSEE    S6TN00240   
8                               GILLIBRAND FOR SENATE    S0NY00410   
9                              GINGREY FOR SENATE INC    S2GA00100   
10                         MARTIN HEINRICH FOR SENATE    S2NM00088   
11                            FRIENDS OF MAZIE HIRONO    S2HI00106   
12                    HUFFMAN 2018 CAMPAIGN COMMITTEE    H8NC08091   
13                             FRIENDS OF DAVID JOLLY    H4FL13101   
14                               NEBRASKANS FOR KLEEB    S8NE00133   
15                        FAMILIES FOR JAMES LANKFORD    S4OK00232   
16                   ROBERT LAUTEN FOR US SENATE 2012    S2CA00690   
17  FREEDOM SOCIALIST 2012 PRESIDENTIAL CAMPAIGN C...    P20004651   
18                                     PEOPLE FOR BEN    S0NM00058   
19                                    DML FOR AMERICA    P60006814   
20                       CITIZENS FOR JOSH MANDEL INC    H0OH11072   

### Integrity of `cand_cmte` intersect table ###

Check for bad Committee keys (i.e. if `cand_cmte` has any `cmte_id` values that don't exist in the `cmte` table)

In [22]:
%%sql
select count(*)
  from cand_cmte cc
 where not exists
       (select *
          from cmte cm
         where cm.cmte_id = cc.cmte_id)

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


count
0      0

Do the same thing, except add `elect_cycle` to the (anti) join

In [23]:
%%sql
select count(*)
  from cand_cmte cc
 where not exists
       (select *
          from cmte cm
         where cm.cmte_id     = cc.cmte_id
           and cm.elect_cycle = cc.elect_cycle)

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


count
0      0

Check for bad Candidate keys (i.e. if `cand_cmte` has any `cand_id` values that don't exist in the `cand` table)

In [24]:
%%sql
select count(*)
  from cand_cmte cc
 where not exists
       (select *
          from cand cn
         where cn.cand_id = cc.cand_id)

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


count
0     11

Let's take a look at the Committees involved

In [25]:
%%sql
select cm.elect_cycle,
       cm.cmte_id,
       cm.cmte_nm,
       cm.cmte_tp,
       cm.org_tp,
       cm.cmte_pty_affiliation,
       cc.cand_election_yr,
       cc.fec_election_yr
  from cand_cmte cc
  join cmte cm on  cm.cmte_id     = cc.cmte_id
               and cm.elect_cycle = cc.elect_cycle
 where not exists
       (select *
          from cand cn
         where cn.cand_id = cc.cand_id)
 order by 2, 1

 * postgresql+psycopg2://crash@caladan/fecdb
11 rows affected.


elect_cycle    cmte_id                                         cmte_nm  \
0          2000  C00356071  BILL HAAS FOR CONGRESS (1ST DISTRICT MISSOURI)   
1          2002  C00356071  BILL HAAS FOR CONGRESS (1ST DISTRICT MISSOURI)   
2          2000  C00363937                     MIKE GALLAGHER FOR CONGRESS   
3          2016  C00625046                COMMITTEE TO ELECT BENJAMIN NOFS   
4          2018  C00625046                COMMITTEE TO ELECT BENJAMIN NOFS   
5          2016  C00626838                     SHELLY SCHRATZ FOR CONGRESS   
6          2018  C00650424                       DWIGHT BRADY FOR CONGRESS   
7          2018  C00677690               WASHINGTON FOR CONGRESS COMMITTEE   
8          2020  C00677690               WASHINGTON FOR CONGRESS COMMITTEE   
9          2020  C00712851                       AMERICAN RELIGIOUS  PARTY   
10         2020  C00716191                        LOGAN WILDE FOR CONGRESS   

   cmte_tp org_tp cmte_pty_affiliation  cand_election_yr  fec_election_yr  
0        H   None                  DEM              2000             2000  
1        H   None                  DEM              2000             2002  
2        H   None                  DEM              2002             2000  
3        H   None                  GRN              2016             2016  
4        H   None                  GRN              2016             2018  
5        H   None                 None              2016             2016  
6        H   None                  DEM              2018             2018  
7        H   None                 None              2018             2018  
8        H   None                 None              2018             2020  
9        X   None                  UNK              2019             2020  
10       H   None                  REP              2020             2020

Let's check for bad Candidate keys again, except this time we'll add `elect_cycle` to the (anti) join

In [26]:
%%sql
select count(*)
  from cand_cmte cc
 where not exists
       (select *
          from cand cn
         where cn.cand_id     = cc.cand_id
           and cn.elect_cycle = cc.elect_cycle)

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


count
0     18

And now take a look at the Committees involved

In [27]:
%%sql
select cm.elect_cycle,
       cm.cmte_id,
       cm.cmte_nm,
       cm.cmte_tp,
       cm.org_tp,
       cm.cmte_pty_affiliation,
       cc.cand_election_yr,
       cc.fec_election_yr
  from cand_cmte cc
  join cmte cm on  cm.cmte_id     = cc.cmte_id
               and cm.elect_cycle = cc.elect_cycle
 where not exists
       (select *
          from cand cn
         where cn.cand_id     = cc.cand_id
           and cn.elect_cycle = cc.elect_cycle)
 order by 2, 1

 * postgresql+psycopg2://crash@caladan/fecdb
18 rows affected.


elect_cycle    cmte_id                                         cmte_nm  \
0          2000  C00330043                   ROSKAM FOR CONGRESS COMMITTEE   
1          2000  C00342261                  PEOPLE FOR AMERICAN LEADERSHIP   
2          2000  C00356071  BILL HAAS FOR CONGRESS (1ST DISTRICT MISSOURI)   
3          2002  C00356071  BILL HAAS FOR CONGRESS (1ST DISTRICT MISSOURI)   
4          2000  C00363937                     MIKE GALLAGHER FOR CONGRESS   
5          2004  C00390468                DRAFT HILLARY FOR PRESIDENT 2004   
6          2004  C00400473              HILLARY CLINTON FOR PRESIDENT 2008   
7          2012  C00459768                JOSE RUIZ FOR CONGRESS COMMITTEE   
8          2016  C00625046                COMMITTEE TO ELECT BENJAMIN NOFS   
9          2018  C00625046                COMMITTEE TO ELECT BENJAMIN NOFS   
10         2016  C00626838                     SHELLY SCHRATZ FOR CONGRESS   
11         2018  C00636308                   GABRIEL MCARTHUR FOR CONGRESS   
12         2020  C00636308                   GABRIEL MCARTHUR FOR CONGRESS   
13         2018  C00650424                       DWIGHT BRADY FOR CONGRESS   
14         2018  C00677690               WASHINGTON FOR CONGRESS COMMITTEE   
15         2020  C00677690               WASHINGTON FOR CONGRESS COMMITTEE   
16         2020  C00712851                       AMERICAN RELIGIOUS  PARTY   
17         2020  C00716191                        LOGAN WILDE FOR CONGRESS   

   cmte_tp org_tp cmte_pty_affiliation  cand_election_yr  fec_election_yr  
0        H   None                  REP              1998             2000  
1        U   None                 None              2000             2000  
2        H   None                  DEM              2000             2000  
3        H   None                  DEM              2000             2002  
4        H   None                  DEM              2002             2000  
5        U   None                  DEM              2004             2004  
6        U   None                  DEM              2004             2004  
7        H   None                  DEM              2012             2012  
8        H   None                  GRN              2016             2016  
9        H   None                  GRN              2016             2018  
10       H   None                 None              2016             2016  
11       H   None                  DEM              2018             2018  
12       H   None                  DEM              2018             2020  
13       H   None                  DEM              2018             2018  
14       H   None                 None              2018             2018  
15       H   None                 None              2018             2020  
16       X   None                  UNK              2019             2020  
17       H   None                  REP              2020             2020

Let's see how many `cand_cmte` records don't join to both `cmte` and `cand`

In [28]:
%%sql
with bad_cand_cmte as (
    select cc.*
      from cand_cmte cc
      left join cmte cm on cm.cmte_id = cc.cmte_id
      left join cand cn on cn.cand_id = cc.cand_id
     where cm.id is null
        or cn.id is null
)
select count(*)
  from bad_cand_cmte

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


count
0     18

And let's do the same thing, adding `elect_cycle` to the joins (this is purely academic, since it looks like the previous result replicates bad Candidate keys by election cycle examination, above&mdash;this query should return the same as the previous one)

In [29]:
%%sql
with bad_cand_cmte as (
    select cc.*
      from cand_cmte cc
      left join cmte cm on  cm.cmte_id     = cc.cmte_id
                        and cm.elect_cycle = cc.elect_cycle
      left join cand cn on  cn.cand_id     = cc.cand_id
                        and cn.elect_cycle = cc.elect_cycle
     where cm.id is null
        or cn.id is null
)
select count(*)
  from bad_cand_cmte

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


count
0     18

Now let's check to see if there are duplicate entries in `cand_cmte`, meaning identical values for `cmte_id`, `cand_id`, and `elect_cycle`

In [30]:
%%sql
with cand_cmte_dup as (
    select cmte_id,
           cand_id,
           elect_cycle,
           count(*) dups,
           array_agg(id)
      from cand_cmte
     group by 1, 2, 3
    having count(*) > 1
)
select elect_cycle,
       dups,
       count(*)
  from cand_cmte_dup
 group by 1, 2
 order by 1, 2 desc

 * postgresql+psycopg2://crash@caladan/fecdb
8 rows affected.


elect_cycle  dups  count
0         2000     2      1
1         2008     2     23
2         2010     2     40
3         2012     2     32
4         2014     2     53
5         2016     2     14
6         2018     2     60
7         2020     2     28

Note that there may be a legitimate reason for these duplications; let's see if/how these records differ from each other in columns other than `cmte_id`, `cand_id`, and `elect_cycle`.  We'll take the first 25 examples as a sampling to inspect...

In [31]:
%%sql
with cand_cmte_dup as (
    select cmte_id,
           cand_id,
           elect_cycle,
           count(*) dups,
           array_agg(id) as cand_cmte_pks
      from cand_cmte
     group by 1, 2, 3
    having count(*) > 1
     limit 25
)
select *
  from cand_cmte
 where id in
       (select unnest(cand_cmte_pks)
          from cand_cmte_dup)
 order by cand_id, cmte_id, elect_cycle

 * postgresql+psycopg2://crash@caladan/fecdb
50 rows affected.


id    cand_id  cand_election_yr  fec_election_yr    cmte_id cmte_tp  \
0   30928  H0CA10099              2010             2010  C00461103       H   
1   30929  H0CA10099              2009             2010  C00461103       H   
2   30937  H0CA10131              2009             2010  C00461947       H   
3   30938  H0CA10131              2010             2010  C00461947       H   
4   30943  H0CA10172              2010             2010  C00463257       H   
5   30944  H0CA10172              2009             2010  C00463257       H   
6   30969  H0CA15171              2009             2010  C00458117       H   
7   30970  H0CA15171              2010             2010  C00458117       H   
8     211  H0FL09107              2019             2020  C00709188       H   
9     212  H0FL09107              2020             2020  C00709188       H   
10  31404  H0IL05088              2009             2010  C00457630       H   
11  31405  H0IL05088              2010             2010  C00457630       H   
12  25732  H0NY09072              2012             2012  C00483495       H   
13  25733  H0NY09072              2011             2012  C00483495       H   
14  32043  H0NY23065              2009             2010  C00463976       H   
15  32044  H0NY23065              2010             2010  C00463976       H   
16  19894  H2IL02172              2013             2014  C00539866       H   
17  19893  H2IL02172              2014             2014  C00539866       H   
18  21265  H4LA05080              2013             2014  C00548503       H   
19  21266  H4LA05080              2014             2014  C00548503       H   
20  21316  H4MA05134              2013             2014  C00548560       H   
21  21317  H4MA05134              2014             2014  C00548560       H   
22  21442  H4MO08188              2013             2014  C00542944       H   
23  21443  H4MO08188              2014             2014  C00542944       H   
24   6692  H6GA06116              2018             2018  C00631168       H   
25   6693  H6GA06116              2017             2018  C00631168       H   
26  15358  H6OH08166              2015             2016  C00588863       H   
27  15359  H6OH08166              2016             2016  C00588863       H   
28   7560  H8CA34191              2018             2018  C00630913       H   
29   7561  H8CA34191              2017             2018  C00630913       H   
30   7572  H8CA34324              2018             2018  C00635409       H   
31   7573  H8CA34324              2017             2018  C00635409       H   
32  22382  H8CA41063              2014             2014  C00511519       H   
33  22377  H8CA41063              2013             2014  C00511519       H   
34  22386  H8CA41063              2013             2014  C00546382       H   
35  22384  H8CA41063              2014             2014  C00546382       H   
36  38054  H8GA04075              2007             2008  C00433722       H   
37  38053  H8GA04075              2008             2008  C00433722       H   
38  38076  H8GA10072              2008             2008  C00434449       H   
39  38075  H8GA10072              2007             2008  C00434449       H   
40  22497  H8IL11113              2014             2014  C00452045       N   
41  22495  H8IL11113              2013             2014  C00452045       N   
42   9566  H8VA10155              2018             2018  C00651224       H   
43   9567  H8VA10155              2017             2018  C00651224       H   
44  35178  S0MA00125              2009             2010  C00467670       S   
45  35179  S0MA00125              2010             2010  C00467670       S   
46  23981  S4MA00028              2014             2014  C00545210       N   
47  23977  S4MA00028              2013             2014  C00545210       N   
48  23994  S4MA00218              2013             2014  C00542142       S   
49  23992  S4MA00218              2014             2014  C00542142       S   

   cmte_dsgn linkage_id  ele

Looks like it may be due to `cand_election_yr`, so let's factor that into the equation and see to what extent that is true

In [32]:
%%sql
with cand_cmte_dup as (
    select cmte_id,
           cand_id,
           cand_election_yr,
           elect_cycle,
           count(*) dups,
           array_agg(id)
      from cand_cmte
     group by 1, 2, 3, 4
    having count(*) > 1
)
select elect_cycle,
       dups,
       count(*)
  from cand_cmte_dup
 group by 1, 2
 order by 1, 2 desc

 * postgresql+psycopg2://crash@caladan/fecdb
2 rows affected.


elect_cycle  dups  count
0         2018     2      2
1         2020     2      2

And let's see what the remaining exceptions are

In [33]:
%%sql
with cand_cmte_dup as (
    select cmte_id,
           cand_id,
           cand_election_yr,
           elect_cycle,
           count(*) dups,
           array_agg(id) as cand_cmte_pks
      from cand_cmte
     group by 1, 2, 3, 4
    having count(*) > 1
     limit 25
)
select *
  from cand_cmte
 where id in
       (select unnest(cand_cmte_pks)
          from cand_cmte_dup)
 order by cand_id, cmte_id, elect_cycle

 * postgresql+psycopg2://crash@caladan/fecdb
8 rows affected.


id    cand_id  cand_election_yr  fec_election_yr    cmte_id cmte_tp  \
0  7927  H8GA04117              2018             2018  C00654699       H   
1  7926  H8GA04117              2018             2018  C00654699       H   
2  2213  H8GA04117              2018             2020  C00654699       H   
3  2212  H8GA04117              2018             2020  C00654699       H   
4  8672  H8NJ07256              2018             2018  C00690123       O   
5  8673  H8NJ07256              2018             2018  C00690123       O   
6  2698  H8NJ07256              2018             2020  C00690123       O   
7  2697  H8NJ07256              2018             2020  C00690123       O   

  cmte_dsgn linkage_id  elect_cycle  
0         P     217917         2018  
1         P     217918         2018  
2         P     223620         2020  
3         P     223619         2020  
4         U     221986         2018  
5         U     221985         2018  
6         U     224070         2020  
7         U     224071         2020

### Consistency of Committees to Candidates, through `cmte.cand_id` foreign key ###

We first check `cmte` for `cmte_id`'s with different `cand_id`'s within any election cycle

In [34]:
%%sql
with ec_cmte_cand_diffs as (
    select elect_cycle,
           cmte_id,
           count(distinct cand_id) as distinct_cand_ids,
           array_agg(distinct cand_id) as cand_ids
      from cmte
     group by 1, 2
    having count(distinct cand_id) > 1
)
select elect_cycle,
       distinct_cand_ids,
       count(*) as ec_cmte_ids
  from ec_cmte_cand_diffs
 group by 1, 2
 order by 1, 2 desc

 * postgresql+psycopg2://crash@caladan/fecdb
0 rows affected.


Empty DataFrame
Columns: []
Index: []

If not, now let's check for `cmte_id`'s with different `cand_id`'s *across* election cycles

In [35]:
%%sql
with cmte_cand_diffs as (
    select cmte_id,
           count(distinct cand_id) as distinct_cand_ids,
           array_agg(distinct cand_id) as cand_ids
      from cmte
     group by 1
    having count(distinct cand_id) > 1
)
select distinct_cand_ids,
       count(*) as cmte_ids
  from cmte_cand_diffs
 group by 1
 order by 1 desc

 * postgresql+psycopg2://crash@caladan/fecdb
2 rows affected.


distinct_cand_ids  cmte_ids
0                  3         3
1                  2       228

Let's captures some of the instances, so we can visually inspect the disparate Candidate information for a specific Committee (as identified by `cmte_id`)

In [36]:
%%sql result <<
with cmte_cand_diffs as (
    select cmte_id,
           count(distinct cand_id) as distinct_cand_ids,
           array_agg(distinct cand_id) as cand_ids
      from cmte
     group by 1
    having count(distinct cand_id) > 1
)
select *
  from cmte_cand_diffs
 order by 2 desc
 limit 50

 * postgresql+psycopg2://crash@caladan/fecdb
50 rows affected.
Returning data to local variable result


We'll take a look at the first ten instances...

In [37]:
cmte_id0 = result.loc[0][0]
cmte_id1 = result.loc[1][0]
cmte_id2 = result.loc[2][0]
cmte_id3 = result.loc[3][0]
cmte_id4 = result.loc[4][0]
cmte_id5 = result.loc[5][0]
cmte_id6 = result.loc[6][0]
cmte_id7 = result.loc[7][0]
cmte_id8 = result.loc[8][0]
cmte_id9 = result.loc[9][0]
result[0:9]

cmte_id  distinct_cand_ids                           cand_ids
0  C00354597                  3  [H0NJ10117, H2FL11075, P40003345]
1  C00501197                  3  [H2TX16185, P00010793, S8TX00285]
2  C00567677                  3  [H0NY19139, H2AZ08102, H4CA45097]
3  C00254938                  2       [H2DE00072, S0DE00068, None]
4  C00266932                  2             [H2GA08038, S2GA00118]
5  C00267708                  2             [H2FL16041, S4FL00207]
6  C00264697                  2             [H2OH13033, S6OH00163]
7  C00303552                  2             [H2NJ06053, S6NJ00164]
8  C00304758                  2             [H6MO07128, S0MO00183]

We sort by `cmte_id`, so the variation in Candidate ID and name will be in adjacent rows.  It is interesting to see whether `cand_id` just switches at a point in time (election cycle), or whether multiple values are used haphazardly between cycles

In [38]:
%%sql
select cmte.cmte_id,
       cand.cand_id,
       cand.cand_name,
       array_agg(distinct cand.elect_cycle) as elect_cycles
  from cmte
  join cand on  cand.cand_id     = cmte.cand_id
            and cand.elect_cycle = cmte.elect_cycle
 where cmte.cmte_id in (:cmte_id0, :cmte_id1, :cmte_id2, :cmte_id3, :cmte_id4,
                        :cmte_id5, :cmte_id6, :cmte_id7, :cmte_id8, :cmte_id9)
 group by 1, 2, 3
 order by 1, 2, 3

 * postgresql+psycopg2://crash@caladan/fecdb
28 rows affected.


cmte_id    cand_id                  cand_name  \
0   C00254938  H2DE00072          CASTLE, MICHAEL N   
1   C00254938  S0DE00068          CASTLE, MICHAEL N   
2   C00264697  H2OH13033             BROWN, SHERROD   
3   C00264697  S6OH00163             BROWN, SHERROD   
4   C00266932  H2GA08038           CHAMBLISS, SAXBY   
5   C00266932  S2GA00118         CHAMBLISS, C SAXBY   
6   C00266932  S2GA00118           CHAMBLISS, SAXBY   
7   C00267708  H2FL16041           DEUTSCH, PETER R   
8   C00267708  S4FL00207             DEUTSCH, PETER   
9   C00303552  H2NJ06053            SMITH, ROBERT G   
10  C00303552  S6NJ00164            SMITH, ROBERT G   
11  C00304758  H6MO07128                 BLUNT, ROY   
12  C00304758  S0MO00183                 BLUNT, ROY   
13  C00307165  H6SD00085              THUNE, JOHN R   
14  C00307165  S2SD00068                THUNE, JOHN   
15  C00307165  S2SD00068              THUNE, JOHN R   
16  C00354597  H0NJ10117               FLASH, ROCKY   
17  C00354597  H2FL11075               FLASH, ROCKY   
18  C00354597  P40003345     FLASH, ROCKY HURRICANE   
19  C00501197  H2TX16185      O'ROURKE, ROBERT BETO   
20  C00501197  H2TX16185  O'ROURKE, ROBERT ''BETO''   
21  C00501197  H2TX16185    O'ROURKE, ROBERT (BETO)   
22  C00501197  P00010793      O'ROURKE, ROBERT BETO   
23  C00501197  S8TX00285    O'ROURKE, ROBERT (BETO)   
24  C00567677  H0NY19139              HAYWORTH, NAN   
25  C00567677  H2AZ08102            MCSALLY, MARTHA   
26  C00567677  H2AZ08102     MCSALLY, MARTHA E. MS.   
27  C00567677  H4CA45097              WALTERS, MIMI   

                                        elect_cycles  
0                     [2000, 2002, 2004, 2006, 2008]  
1                                       [2010, 2012]  
2                                 [2000, 2002, 2004]  
3   [2006, 2008, 2010, 2012, 2014, 2016, 2018, 2020]  
4                                             [2000]  
5                     [2012, 2014, 2016, 2018, 2020]  
6                     [2002, 2004, 2006, 2008, 2010]  
7                                       [2000, 2002]  
8                                 [2004, 2006, 2008]  
9                                       [2002, 2004]  
10                                            [2000]  
11                    [2000, 2002, 2004, 2006, 2008]  
12              [2010, 2012, 2014, 2016, 2018, 2020]  
13                                            [2000]  
14                                            [2002]  
15                                            [2004]  
16                                            [2000]  
17                                            [2002]  
18                                      [2004, 2006]  
19                                            [2016]  
20                                            [2012]  
21                                            [2014]  
22                                            [2020]  
23                                            [2018]  
24                                            [2014]  
25                                            [2018]  
26                                            [2016]  
27                                            [2020]

### Cardinality of Committees to Candidates, through `cand_cmte` intersect table ###

See how many `cmte` records join (or attempt to join) to `cand` through `cand_cmte`

In [39]:
%%sql
select count(*) as link_recs,
       round(count(*)::numeric / :cmte_count_total * 100.0, 2) as pct_link_recs,
       count(cn.id) as cand_recs,
       round(count(cn.id)::numeric / :cmte_count_total * 100.0, 2) as pct_cand_recs
  from cmte cm
  join cand_cmte cc on  cc.cmte_id     = cm.cmte_id
                    and cc.elect_cycle = cm.elect_cycle
  left join cand cn on  cn.cand_id     = cc.cand_id
                    and cn.elect_cycle = cc.elect_cycle

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


link_recs pct_link_recs  cand_recs pct_cand_recs
0      55899         40.39      55881         40.38

In [40]:
%%sql
select count(*) as link_recs,
       round(count(*)::numeric / :cmte_count_total * 100.0, 2) as pct_link_recs,
       count(cn.id) as cand_recs,
       round(count(cn.id)::numeric / :cmte_count_total * 100.0, 2) as pct_cand_recs
  from cmte cm
  join cand_cmte cc on  cc.cmte_id     = cm.cmte_id
                    and cc.elect_cycle = cm.elect_cycle
  left join cand cn on  cn.cand_id     = cc.cand_id
                    and cn.elect_cycle = cc.elect_cycle

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


link_recs pct_link_recs  cand_recs pct_cand_recs
0      55899         40.39      55881         40.38

See how many `cmte` records join to multiple `cand_cmte` ("link") and `cand` records, and in what cardinality; show the `cmte.id` ("pk") values with the highest number of link/Candidate associations

In [41]:
%%sql
with multi_cand_cmte as (
    select cm.id as cmte_pk,
           count(*) as link_recs,
           count(cn.id) as cand_recs,
           array_agg(distinct cn.cand_id) as cand_ids
      from cmte cm
      join cand_cmte cc on  cc.cmte_id     = cm.cmte_id
                        and cc.elect_cycle = cm.elect_cycle
      left join cand cn on  cn.cand_id     = cc.cand_id
                        and cn.elect_cycle = cc.elect_cycle
     group by 1
    having count(*) > 1
)
select link_recs,
       cand_recs,
       count(*)           as num_cmtes,
       array_agg(cmte_pk) as cmte_pks
  from multi_cand_cmte
 group by 1, 2
 order by 1 desc, 2 desc, 3 desc

 * postgresql+psycopg2://crash@caladan/fecdb
24 rows affected.


link_recs  cand_recs  num_cmtes  \
0          43         43          2   
1          31         31          3   
2          22         22          8   
3          21         21          2   
4          20         20         11   
5          19         19          5   
6          18         18          3   
7          17         17          1   
8          16         16         11   
9          15         15          4   
10         14         14         14   
11         13         13          5   
12         12         12         13   
13         11         11         13   
14         10         10         25   
15          9          9         20   
16          8          8         30   
17          7          7         23   
18          6          6         38   
19          5          5         77   
20          4          4        158   
21          3          3        180   
22          2          2        971   
23          2          1          3   

                                             cmte_pks  
0                                      [57967, 74923]  
1                              [71683, 87658, 131737]  
2   [95639, 106101, 107390, 107653, 116269, 117119...  
3                                      [61867, 87328]  
4   [73669, 83802, 86615, 90337, 95512, 99832, 998...  
5                 [43255, 58784, 63241, 76005, 98290]  
6                             [87582, 106839, 136132]  
7                                            [109356]  
8   [4073, 11334, 18671, 25086, 38725, 57412, 7413...  
9                        [42135, 59902, 63442, 77472]  
10  [4633, 19694, 40260, 42436, 57505, 60015, 6380...  
11                 [9865, 29836, 73296, 74040, 89870]  
12  [10218, 21389, 30266, 42984, 55533, 58334, 593...  
13  [25279, 42796, 58690, 61479, 63601, 63663, 710...  
14  [6462, 24441, 48055, 59560, 61863, 63420, 7446...  
15  [24514, 48145, 58004, 59714, 60108, 61572, 637...  
16  [5314, 21057, 24593, 42449, 47064, 48239, 5967...  
17  [7557, 23938, 26596, 42831, 47433, 47573, 4816...  
18  [28110, 42330, 42500, 42855, 43180, 44729, 466...  
19  [21388, 21403, 23251, 25757, 42983, 43008, 465...  
20  [3873, 8885, 9563, 9709, 10762, 18346, 23358, ...  
21  [8953, 8954, 9089, 22221, 25322, 26734, 26946,...  
22  [3341, 4075, 4384, 5098, 5152, 5215, 5263, 534...  
23                               [6939, 25509, 26434]

Now let's take a look at the actual Committee records at the top of the list, and validate that the `cand_id` for the record (if specified) is actually in the `cand_cmte` link assocation as well

In [42]:
%%sql
with multi_cand_cmte as (
    select cm.id as cmte_pk,
           count(*) as link_recs,
           count(cn.id) as cand_recs,
           array_agg(distinct cn.cand_id) as cand_ids
      from cmte cm
      join cand_cmte cc on  cc.cmte_id     = cm.cmte_id
                        and cc.elect_cycle = cm.elect_cycle
      left join cand cn on  cn.cand_id     = cc.cand_id
                        and cn.elect_cycle = cc.elect_cycle
     group by 1
    having count(*) > 1
)
select cm.id as cmte_pk,
       cm.elect_cycle,
       cm.cmte_id,
       cm.cmte_nm,
       cm.cand_id,
       mcc.link_recs,
       cm.cand_id = any(mcc.cand_ids) as cand_id_link_rec
  from multi_cand_cmte mcc
  join cmte cm on cm.id = mcc.cmte_pk
 order by 6 desc
 limit 50

 * postgresql+psycopg2://crash@caladan/fecdb
50 rows affected.


cmte_pk  elect_cycle    cmte_id  \
0     74923         2012  C00501825   
1     57967         2014  C00501825   
2     71683         2012  C00461913   
3    131737         2000  C00242941   
4     87658         2010  C00461913   
5    127679         2002  C00376608   
6    117119         2004  C00386342   
7    119032         2004  C00405977   
8    106101         2006  C00386342   
9     95639         2008  C00396226   
10   116269         2004  C00376608   
11   107390         2006  C00402958   
12   107653         2006  C00405977   
13    87328         2010  C00458562   
14    61867         2014  C00547877   
15   135670         2000  C00340141   
16   106597         2006  C00393181   
17    99892         2008  C00450460   
18    83802         2010  C00393181   
19    86615         2010  C00450460   
20    73669         2012  C00488874   
21    90337         2010  C00488874   
22   137232         2000  C00356287   
23   117781         2004  C00393181   
24    95512         2008  C00393181   
25    99832         2008  C00449850   
26    76005         2012  C00512749   
27    58784         2014  C00512749   
28    43255         2016  C00576280   
29    63241         2014  C00561845   
30    98290         2008  C00434274   
31    87582         2010  C00461152   
32   136132         2000  C00344952   
33   106839         2006  C00396226   
34   109356         2006  C00423665   
35    57412         2014  C00493783   
36     4073         2020  C00493783   
37    74130         2012  C00493783   
38   126626         2002  C00365585   
39    25086         2018  C00631960   
40   118559         2004  C00401158   
41    18671         2018  C00493783   
42    11334         2020  C00701243   
43    97348         2008  C00423590   
44   109349         2006  C00423590   
45    38725         2016  C00493783   
46    77472         2012  C00527549   
47    59902         2014  C00527549   
48    63442         2014  C00563858   
49    42135         2016  C00563858   

                                              cmte_nm    cand_id  link_recs  \
0                       JARED POLIS VICTORY FUND 2012  S2IN00091         43   
1                       JARED POLIS VICTORY FUND 2012  S2IN00091         43   
2                            JARED POLIS VICTORY FUND  S0CO00211         31   
3             REPUBLICAN SENATORIAL INNER CIRCLE 1990  S0AK00063         31   
4                            JARED POLIS VICTORY FUND  S0CO00211         31   
5                                  W/N 2002 C0MMITTEE  S0IA00077         22   
6                                 FRONTLINE DEMOCRATS  H0GA08032         22   
7                   2004 JOINT CANDIDATE COMMITTEE II  S2SD00068         22   
8                                 FRONTLINE DEMOCRATS  H0GA08032         22   
9                           SENATE MAJORITY COMMITTEE  S4AZ00030         22   
10                                 W/N 2002 C0MMITTEE  S0ID00057         22   
11                      COMMITTEE FOR A BETTER FUTURE  S4CO00163         22   
12                  2004 JOINT CANDIDATE COMMITTEE II  S2SD00068         22   
13                     HOUSE REPUBLICAN FRESHMEN 2009  H6NJ12136         21   
14                                  TEXANS FOR TEXANS  H0TX04037         21   
15                   NEW CONSERVATIVE LEADERSHIP FUND  H4IN09049         20   
16                     WESTERN GOP MAJORITY COMMITTEE  H0CA14042         20   
17                      2008 JOINT CANDIDATE COMMITTE  S0NH00201         20   
18                     WESTERN GOP MAJORITY COMMITTEE  H0CA48024         20   
19                      2008 JOINT CANDIDATE COMMITTE  S6ME00159         20   
20                 FOUNDERS JOINT CANDIDATE COMMITTEE  S0FL00338         20   
21                 FOUNDERS JOINT CANDIDATE COMMITTEE  S0CA00330         20   
22                                 W/N 2000 COMMITTEE  S8VA00107         20   
23                     WESTERN GOP MAJORITY COMMITTEE  H0CA14042         20   
24                     WESTERN GOP MAJO

See if there are any `cmte` records with multiple `cand` associations through the link table but no direct join through `cmte.cand_id`

In [43]:
%%sql
with multi_cand_cmte as (
    select cm.id as cmte_pk,
           count(*) as link_recs,
           count(cn.id) as cand_recs,
           array_agg(distinct cn.cand_id) as cand_ids
      from cmte cm
      join cand_cmte cc on  cc.cmte_id     = cm.cmte_id
                        and cc.elect_cycle = cm.elect_cycle
      left join cand cn on  cn.cand_id     = cc.cand_id
                        and cn.elect_cycle = cc.elect_cycle
     group by 1
    having count(*) > 1
)
select mcc.link_recs,
       mcc.cand_recs,
       count(*),
       array_agg(mcc.cmte_pk) as cmte_pks
  from multi_cand_cmte mcc
  join cmte cm on cm.id = mcc.cmte_pk
 where cm.cand_id is null
 group by 1, 2
 order by 1 desc, 2 desc, 3 desc

 * postgresql+psycopg2://crash@caladan/fecdb
0 rows affected.


Empty DataFrame
Columns: []
Index: []

See if there are any `cmte` records with a single `cand` association through the link table but a mismatching direct join through `cmte.cand_id`

In [44]:
%%sql
with multi_cand_cmte as (
    select cm.id as cmte_pk,
           count(*) as link_recs,
           sum((cm.cand_id != cc.cand_id)::integer) as mismatches
      from cmte cm
      join cand_cmte cc on  cc.cmte_id     = cm.cmte_id
                        and cc.elect_cycle = cm.elect_cycle
     group by 1
    having count(*) = 1
)
select count(*) as link_recs,
       sum(mcc.mismatches) as mismatches,
       round(sum(mcc.mismatches)::numeric / count(*) * 100.0, 2) as pct_mismatch
  from multi_cand_cmte mcc

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


link_recs mismatches pct_mismatch
0      49756         12         0.02

And look for any direct Committee-Candidate joins (using `cmte.cand_id`) that don't have a corresponding association through `cand_cmte`)

In [45]:
%%sql
with cmte_cand_join as (
    select cm.*
      from cmte cm
      join cand cn           on  cn.cand_id     = cm.cand_id
                             and cn.elect_cycle = cm.elect_cycle
      left join cand_cmte cc on  cc.cmte_id     = cm.cmte_id
                             and cc.cand_id     = cm.cand_id
                             and cc.elect_cycle = cm.elect_cycle
     where cc.id is null
)
select count(*)
  from cmte_cand_join

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


count
0     56

List them in descending chronological order

In [46]:
%%sql
with cmte_cand_join as (
    select cm.*,
           cn.cand_name,
           cn.cand_pty_affiliation,
           cn.cand_election_yr
      from cmte cm
      join cand cn           on  cn.cand_id     = cm.cand_id
                             and cn.elect_cycle = cm.elect_cycle
      left join cand_cmte cc on  cc.cmte_id     = cm.cmte_id
                             and cc.cand_id     = cm.cand_id
                             and cc.elect_cycle = cm.elect_cycle
     where cc.id is null
)
select elect_cycle,
       cmte_id,
       cmte_nm,
       cand_name,
       cand_pty_affiliation,
       cand_election_yr
  from cmte_cand_join
 order by 1 desc, 3
 limit 100

 * postgresql+psycopg2://crash@caladan/fecdb
56 rows affected.


elect_cycle    cmte_id                                            cmte_nm  \
0          2020  C00709600                   ANAMO (AMERICA NEEDS A MAKEOVER)   
1          2020  C00693267    BLACK LABEL EMPIRE (HOUSE OF LORDS) CYBER UNITS   
2          2020  C00583567                 COMMITTEE TO ELECT ROBERT MARSHALL   
3          2020  C00573105  COMMITTEE TO INSURE CIVIL & CONSTITUTIONAL RIGHTS   
4          2020  C00610022        FINANCE COMMITTEE FOR DOUGLAS HOWARD PIERCE   
5          2020  C00664243                   JAZMINA SAAVEDRA FOR US CONGRESS   
6          2020  C00550863                   MARIANNE WILLIAMSON FOR CONGRESS   
7          2020  C00684381                            M HUDSON HALE COMMITTEE   
8          2020  C00612051                NEW FUTURE FOR MATTHEW KALUS THALER   
9          2020  C00716381                          PATRONS OF ROBERT E SMITH   
10         2020  C00694448   PLATINUM BOSS COMMITTEE OF GURPINDER SINGH SAGOO   
11         2020  C00678938                        PRESIDENT DONNA JEAN ALSTON   
12         2020  C00677187         STEPHEN J. SCHRADER FOR U.S. CONGRESS 2020   
13         2020  C00694950  THE FRIENDS OF JAMES JEROME BELL FOR PRESIDENT...   
14         2020  C00514265                        WENDY LONG FOR NEW YORK INC   
15         2018  C00573105  COMMITTEE TO INSURE CIVIL & CONSTITUTIONAL RIGHTS   
16         2018  C00663617                                          COPETEXAS   
17         2018  C00445445                               DAVIS FOR THE FUTURE   
18         2018  C00640193                               ELOHO YISAH BANK PLC   
19         2018  C00664300                                        EMILY ROSAS   
20         2018  C00640326                             GODWIN OYEDEPO KINGDOM   
21         2018  C00432740                               INTERIGATED REVIVAIL   
22         2018  C00664243                   JAZMINA SAAVEDRA FOR US CONGRESS   
23         2018  C00663906                                 KEVIN FOR CONGRESS   
24         2018  C00684381                            M HUDSON HALE COMMITTEE   
25         2018  C00612051                NEW FUTURE FOR MATTHEW KALUS THALER   
26         2018  C00690248                           PEOPLE FOR IAN SCHLAKMAN   
27         2018  C00621110                                PEOPLE WITH CHARLIE   
28         2018  C00678938                        PRESIDENT DONNA JEAN ALSTON   
29         2018  C00588426                                      ROCKY 101 LLC   
30         2018  C00677187         STEPHEN J. SCHRADER FOR U.S. CONGRESS 2020   
31         2018  C00638908                                        TALEM GOMEZ   
32         2018  C00654095                             TARKANIAN FOR CONGRESS   
33         2018  C00434522                                 VAUGHN 4 US SENATE   
34         2018  C00514265                        WENDY LONG FOR NEW YORK INC   
35         2016  C00517482                               BARRON FOR US SENATE   
36         2016  C00573105  COMMITTEE TO INSURE CIVIL & CONSTITUTIONAL RIGHTS   
37         2016  C00541904                                JOE CARR FOR SENATE   
38         2016  C00612051                NEW FUTURE FOR MATTHEW KALUS THALER   
39         2016  C00630327                STRICKLAND 2016 FOR PRESIDENT OF US   
40         2016  C00434522                                 VAUGHN 4 US SENATE   
41         2016  C00514265                        WENDY LONG FOR NEW YORK INC   
42         2014  C00497511                             BACHMANN FOR PRESIDENT   
43         2014  C00514935                      COMMITTEE TO ELECT BILL FLYNN   
44         2014  C00552505             COMMITTEE TO ELECT MICHAEL BICKELMEYER   
45         2014  C00515213                          TIM D'ANNUNZIO FOR SENATE   
46         2012  C00497511                             BACHMANN FOR PRESIDENT   
47         2012  C00420562                         GUGLIELMI FOR CONGRESS LLC   
48         2012  C00492090         

### Cardinality of Candidates to Committees, through `cmte.cand_id` foreign key ###

First see if there are any Candidate records with no directly associated Committee (through the `cmte.cand_id` foreign key)

In [47]:
%%sql result <<
select count(*) as orphan_count,
       round(count(*)::numeric / :cand_count_total * 100.0, 2) as pct_total_cands,
       count(distinct cn.cand_id) orphans_distinct,
       round(count(distinct cn.cand_id)::numeric / :cand_distinct_ids * 100.0, 2) as pct_distinct_cands
  from cand cn
 where not exists
       (select *
          from cmte cm
         where cm.cand_id = cn.cand_id)

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.
Returning data to local variable result


In [48]:
orphan_count     = int(result.loc[0][0])
orphans_distinct = int(result.loc[0][2])
result

orphan_count pct_total_cands  orphans_distinct pct_distinct_cands
0          9904           17.49              7973              30.38

Let's see if the unassociated nature of these Candidate records has something to do with their "status"

In [49]:
%%sql
select cand_status,
       count(*) as orphan_count,
       round(count(*)::numeric / :orphan_count * 100.0, 2) as pct_total_orphans,
       count(distinct cn.cand_id) orphans_distinct,
       round(count(distinct cn.cand_id)::numeric / :orphans_distinct * 100.0, 2) as pct_distinct_orphans
  from cand cn
 where not exists
       (select *
          from cmte cm
         where cm.cand_id = cn.cand_id)
 group by 1
 order by 1

 * postgresql+psycopg2://crash@caladan/fecdb
5 rows affected.


cand_status  orphan_count pct_total_orphans  orphans_distinct  \
0           A             1              0.01                 1   
1           C            35              0.35                34   
2           F             2              0.02                 2   
3           N          9805             99.00              7916   
4           P            61              0.62                37   

  pct_distinct_orphans  
0                 0.01  
1                 0.43  
2                 0.03  
3                99.29  
4                 0.46

For the ones with `cand_status` = "C" ("statutory" candidate), let's see what their PCCs (principal campaign committees) look like&mdash;in partcular, whether they have `cand_id`'s defined, and if so, how those `cand` records relate to the current orphaned record

In [50]:
%%sql
select cn.elect_cycle,
       cn.cand_id,
       cn.cand_name,
       cn.cand_pcc,
       cm.cmte_nm,
       cm.cand_id as cmte_cand_id,
       cn2.cand_name as cmte_cand_name
  from cand cn
  left join cmte cm  on  cm.cmte_id      = cn.cand_pcc
                     and cm.elect_cycle  = cn.elect_cycle
  left join cand cn2 on  cn2.cand_id     = cm.cand_id
                     and cn2.elect_cycle = cm.elect_cycle
 where not exists
       (select *
          from cmte cm
         where cm.cand_id = cn.cand_id)
   and cn.cand_status = 'C'
 order by cn2.id > 0, cm.id > 0, cn.cand_pcc is null, 1 desc, 3

 * postgresql+psycopg2://crash@caladan/fecdb
35 rows affected.


elect_cycle    cand_id                       cand_name   cand_pcc  \
0          2018  H8MD03116           ADAM DAVIDSON DEMARCO  C00660282   
1          2018  S8MI00356              EPSTEIN, LENA ROSE  C00641498   
2          2018  H8MD02118                     MATORY, LIZ  C00583781   
3          2018  S8PA00338        WRIGHT, THERESA MICHELLE  C00659953   
4          2016  S6FL00376              GRAYSON, ALAN MARK  C00424713   
5          2016  S6FL00350                  JOLLY, DAVID W  C00551572   
6          2016  P60006954           LYNCH, DENNIS MICHAEL  C00576074   
7          2016  H6IL18153                  MELLON, ROBERT  C00582460   
8          2016  S6NY00383                 SPOTORNO, FRANK  C00579599   
9          2014  S4MT00084              STAPLETON, COREY C  C00542068   
10         2014  S4IA00103                    YOUNG, DAVID  C00545616   
11         2014  S4GA11269                 YU, EUGENE CHIN  C00547158   
12         2012  S2NY00267                  CARVIN, JOSEPH  C00515601   
13         2012  P20004669    LOPEZ, CHRISTINA (VICE PRES)  C00512343   
14         2012  S2MA00147       ROBINSON, HERBERT WILLIAM  C00451146   
15         2010  S0CO00302             BARTON, STEVEN KENT  C00468710   
16         2000  H2NY02057                   LAZIO, RICK A  C00323618   
17         2000  H0FL05014                  MCCOLLUM, BILL  C00166041   
18         2018  H8MI01164          BRADY, DWIGHT JOHN MR.  C00650424   
19         2020  S0NC00343            SMITH, ERICA DANETTE       None   
20         2018  S8PA00304           ADDIS , PAUL DAVID MR       None   
21         2018  S8ND00104                   CAMPBELL, TOM       None   
22         2018  H8TX06282  COPE, KENNETH CLIFFORD MR. JR.       None   
23         2018  H8LA06147                   DEWITT, JASON       None   
24         2018  H8PA01286                MALIK, DEAN H MR       None   
25         2018  S8WV00176        WILLIS, THOMAS EDWARD II       None   
26         2016  S6FL00293              DESANTIS, RONALD D       None   
27         2012  S2TX00395            BOONE, DANIEL CURTIS       None   
28         2012  S2FL00250                    HASNER, ADAM       None   
29         2012  S2NJ00395                  LITTLE, ANNA C       None   
30         2012  S2OH00154              WALTERS, RYAN DEAN       None   
31         2010  S8MN00479                 DILLON, DAVID G       None   
32         2008  S8MN00479                 DILLON, DAVID G       None   
33         2004  S4FL00231                     FOLEY, MARK       None   
34         2004  P40003428  LAMARCHE, PATRICIA (VICE-PRES)       None   

                                              cmte_nm cmte_cand_id  \
0                     COMMITTEE TO ELECT ADAM DEMARCO    H8DC03014   
1                                   LENA FOR CONGRESS    H8MI11320   
2                             LIZ MATORY FOR CONGRESS    H6MD08507   
3                         CAMPAIGN FOR THERESA WRIGHT    H8PA07259   
4                     COMMITTEE TO ELECT ALAN GRAYSON    H6FL08213   
5                              FRIENDS OF DAVID JOLLY    H4FL13101   
6                                     DML FOR AMERICA    P60006814   
7                             ROB MELLON FOR CONGRESS    H4IL18109   
8                                SPOTORNO FOR AMERICA    H6NY14244   
9                              WWW.COREYSTAPLETON.COM    H4MT01033   
10                               YOUNG FOR IOWA, INC.    H4IA03115   
11                            EUGENE YU FOR US SENATE    H4GA12051   
12                                        CARVIN 2012    H2NY17139   
13  FREEDOM SOCIALIST 2012 PRESIDENTIAL CAMPAIGN C...    P20004651   
14                      HERB ROBINSON COMMITTEE, INC.    H8MA04047   
15                          STEVE BARTON CAMPAIGN INC    H0CO01102   
16                                     LAZIO 2000 INC    S0NY00303   
17                        BILL MCCOLLUM FOR US SENATE    S8FL00075   
18                          DWIGHT BRADY FOR 

Now we'll do the same thing (looking at Candidate records with no `cmte.cand_id` association), with `elect_cycle` added to the (anti) join

In [51]:
%%sql
select count(*) as orphan_count,
       count(distinct cn.cand_id) orphans_distinct
  from cand cn
 where not exists
       (select *
          from cmte cm
         where cm.cand_id     = cn.cand_id
           and cm.elect_cycle = cn.elect_cycle)

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


orphan_count  orphans_distinct
0         11114              8853

Again, we'll summarize by `cand_status`

In [52]:
%%sql
select cand_status,
       count(*) as orphan_count,
       count(distinct cn.cand_id) orphans_distinct
  from cand cn
 where not exists
       (select *
          from cmte cm
         where cm.cand_id     = cn.cand_id
           and cm.elect_cycle = cn.elect_cycle)
 group by 1
 order by 1

 * postgresql+psycopg2://crash@caladan/fecdb
6 rows affected.


cand_status  orphan_count  orphans_distinct
0           A             1                 1
1           C           112               109
2           F             4                 4
3           N         10595              8537
4           P           400               266
5        None             2                 2

And list the ones with `cand_status` = "C" ("statutory" candidate), along with their PCC information (including associated Candidate for the Committee)

In [53]:
%%sql
select cn.elect_cycle,
       cn.cand_id,
       cn.cand_name,
       cn.cand_pcc,
       cm.cmte_nm,
       cm.cand_id as cmte_cand_id,
       cn2.cand_name as cmte_cand_name
  from cand cn
  left join cmte cm  on  cm.cmte_id      = cn.cand_pcc
                     and cm.elect_cycle  = cn.elect_cycle
  left join cand cn2 on  cn2.cand_id     = cm.cand_id
                     and cn2.elect_cycle = cm.elect_cycle
 where not exists
       (select *
          from cmte cm
         where cm.cand_id     = cn.cand_id
           and cm.elect_cycle = cn.elect_cycle)
   and cn.cand_status = 'C'
 order by cn2.id > 0, cm.id > 0, cn.cand_pcc is null, 1 desc, 3

 * postgresql+psycopg2://crash@caladan/fecdb
112 rows affected.


elect_cycle    cand_id                        cand_name   cand_pcc  \
0           2020  P00007492  BENZEL, JULIANNE ELIZABETH MRS.  C00676320   
1           2018  H8MD03116            ADAM DAVIDSON DEMARCO  C00660282   
2           2018  H2TN06030           BLACKBURN, MARSHA MRS.  C00376939   
3           2018  H0ND01026                CRAMER, KEVIN MR.  C00504704   
4           2018  H2MD06195                  DELANEY, JOHN K  C00508416   
5           2018  S8MI00356               EPSTEIN, LENA ROSE  C00641498   
6           2018  H8NC08109                   HUFFMAN, SCOTT  C00663229   
7           2018  H4WV03070                  JENKINS, EVAN H  C00548271   
8           2018  H8MD02118                      MATORY, LIZ  C00583781   
9           2018  H2TX16185            O'ROURKE, ROBERT BETO  C00501197   
10          2018  H0IN04170          ROKITA, THEODORE EDWARD  C00476192   
11          2018  H4MT00050                  ROSENDALE, MATT  C00548289   
12          2018  H6NV03139                     ROSEN, JACKY  C00606939   
13          2018  H2AZ09019                  SINEMA, KYRSTEN  C00508804   
14          2018  S8PA00338         WRIGHT, THERESA MICHELLE  C00659953   
15          2016  H4LA07029     BOUSTANY, CHARLES W. DR. JR.  C00394866   
16          2016  S6FL00376               GRAYSON, ALAN MARK  C00424713   
17          2016  S6FL00350                   JOLLY, DAVID W  C00551572   
18          2016  P60006954            LYNCH, DENNIS MICHAEL  C00576074   
19          2016  S6IL00185                 MARSHALL, ROBERT  C00583567   
20          2016  H6IL18153                   MELLON, ROBERT  C00582460   
21          2016  H6FL01143                      MILLS, JOHN  C00565366   
22          2016  S6NY00383                  SPOTORNO, FRANK  C00579599   
23          2016  H0IN03198               STUTZMAN, MARLIN A  C00484683   
24          2016  H0IN09070          YOUNG, TODD CHRISTOPHER  C00459255   
25          2014  S0CO00237                  BUCK, KENNETH R  C00461368   
26          2014  H4TN04130                        CARR, JOE  C00541904   
27          2014  H2AR04083                   COTTON, THOMAS  C00499988   
28          2014  H0OK05114         LANKFORD, JAMES PAUL MR.  C00466482   
29          2014  H8MI09068                     PETERS, GARY  C00437889   
30          2014  S4MT00084               STAPLETON, COREY C  C00542068   
31          2014  S4IA00103                     YOUNG, DAVID  C00545616   
32          2014  S4GA11269                  YU, EUGENE CHIN  C00547158   
33          2012  H0MO02148                     AKIN, W TODD  C00343475   
34          2012  H8WI00018                   BALDWIN, TAMMY  C00326801   
35          2012  H8NV01071                 BERKLEY, SHELLEY  C00325738   
36          2012  S2NY00267                   CARVIN, JOSEPH  C00515601   
37          2012  S2MT00096                   DAINES, STEVEN  C00491357   
38          2012  H4IN02101       DONNELLY, JOSEPH SIMON MR.  C00393652   
39          2012  H0AZ01184                  FLAKE, JEFF MR.  C00347260   
40          2012  H8NM01224                 HEINRICH, MARTIN  C00434563   
41          2012  H6HI02251               HIRONO, MAZIE MRS.  C00420760   
42          2012  H8CA47052            LAUTEN, ROBERT EDMUND  C00496554   
43          2012  P20004669     LOPEZ, CHRISTINA (VICE PRES)  C00512343   
44          2012  S2MA00147        ROBINSON, HERBERT WILLIAM  C00451146   
45          2012  S0NV00203                 TARKANIAN, DANNY  C00465963   
46          2012  S0TX00142              WILLIAMS, MICHAEL L  C00457960   
47          2010  H6ME01157                  ALLEN, THOMAS H  C00328245   
48          2010  S0CO00302              BARTON, STEVEN KENT  C00468710   
49          2010  H6MO07128                       BLUNT, ROY  C00304758   
50          2010  H2DE00072                CASTLE, MICHAEL N  C00254938   
51          2010  H6IN08210                  ELLSWORTH, BRAD  C00412346   
52          2010  S0CO00229   

Now let's look at all of the cardinalities of Candidate to Committee (minus orphans); we count the number of Candidates with various numbers of direct associations (through `cmte.cand_id`)

In [54]:
%%sql
with cand_cmte_card as (
    select cn.cand_id,
           count(distinct cm.cmte_id) as num_cmtes
      from cand cn
      left join cmte cm on cm.cand_id = cn.cand_id
     group by 1
    having count(distinct cm.cmte_id) > 0
)
select num_cmtes as assoc_cmtes,
       count(*)  as cand_count,
       round(count(*)::numeric / (:cand_distinct_ids - :orphans_distinct) * 100.0, 2) as pct_non_orphans
    from cand_cmte_card
 group by 1
 order by 1 desc

 * postgresql+psycopg2://crash@caladan/fecdb
28 rows affected.


assoc_cmtes  cand_count pct_non_orphans
0            43           1            0.01
1            39           1            0.01
2            34           1            0.01
3            31           1            0.01
4            30           1            0.01
5            29           2            0.01
6            25           1            0.01
7            24           1            0.01
8            23           1            0.01
9            22           3            0.02
10           18           1            0.01
11           17           1            0.01
12           16           1            0.01
13           15           1            0.01
14           14           5            0.03
15           13           5            0.03
16           12           7            0.04
17           11           9            0.05
18           10          18            0.10
19            9          18            0.10
20            8          13            0.07
21            7          24            0.13
22            6          28            0.15
23            5          72            0.39
24            4         144            0.79
25            3         372            2.04
26            2        1421            7.78
27            1       16117           88.22

Let's see who the candidates are that have 10 or more directly associated Committees, including the office(s) they are running for

In [55]:
%%sql
with cand_cmte_card as (
    select cn.cand_id,
           count(distinct cm.cmte_id) as num_cmtes
      from cand cn
      join cmte cm on cm.cand_id = cn.cand_id
     group by 1
    having count(distinct cm.cmte_id) >= 10
)
select cn.cand_id,
       cn.cand_name,
       array_agg(distinct cn.cand_pty_affiliation) as party,
       ccc.num_cmtes,
       array_agg(distinct concat('''', cn.cand_election_yr, ' ', cn.cand_office_st, ' ', cn.cand_office, ''''))
           as "race (year state office)"
  from cand_cmte_card ccc
  join cand cn on cn.cand_id = ccc.cand_id
 group by 1, 2, 4
 order by 4 desc

 * postgresql+psycopg2://crash@caladan/fecdb
86 rows affected.


cand_id                                        cand_name       party  \
0   P00003335                                   BUSH, GEORGE W       [REP]   
1   S0NH00219                                  SHAHEEN, JEANNE       [DEM]   
2   P80003338                                    OBAMA, BARACK       [DEM]   
3   P80003338                  OBAMA, BARACK / JOSEPH R. BIDEN       [DEM]   
4   P20000527                                     NADER, RALPH  [IND, UNK]   
5   P60002458                                  JEWELL, ROGER H       [REP]   
6   P60002458                              JEWELL, ROGER HENRY  [NNE, REP]   
7   P80000235                                    KERRY, JOHN F       [DEM]   
8   P80002801                                   MCCAIN, JOHN S       [REP]   
9   P80002801                                  MCCAIN, JOHN S.       [REP]   
10  S0CA00199                                FEINSTEIN, DIANNE       [DEM]   
11  S2CA00286                                   BOXER, BARBARA       [DEM]   
12  S6MO00305                                MCCASKILL, CLAIRE       [DEM]   
13  P80000748                                        PAUL, RON       [REP]   
14  P80000912                                         GORE, AL       [DEM]   
15  S0CO00211                                BENNET, MICHAEL F       [DEM]   
16  S0AR00028                                    PRYOR, MARK L       [DEM]   
17  S0AR00028                             PRYOR, MARK LUNSFORD       [DEM]   
18  S6LA00227                                 LANDRIEU, MARY L       [DEM]   
19  S6MN00267                                   KLOBUCHAR, AMY       [DEM]   
20  S6MN00267                                 KLOBUCHAR, AMY J  [DEM, DFL]   
21  P80000805                              BUCHANAN, PATRICK J       [REF]   
22  P60007168                                 SANDERS, BERNARD       [DEM]   
23  S0NY00410                    GILLIBRAND, KIRSTEN ELIZABETH       [DEM]   
24  S2MT00096                                   DAINES, STEVEN       [REP]   
25  S6MT00162                                      TESTER, JON       [DEM]   
26  S6MT00162                                   TESTER, R. JON       [DEM]   
27  S8MN00438                                      FRANKEN, AL  [DEM, DFL]   
28  H0IL10302                       DOLD JR., ROBERT JAMES MR.       [REP]   
29  H0IL10302                            DOLD, ROBERT JAMES JR       [REP]   
30  H0IL10302                         DOLD, ROBERT JAMES MR JR       [REP]   
31  S0FL00338                                     RUBIO, MARCO       [REP]   
32  S0OH00133                                     PORTMAN, ROB       [REP]   
33  S0OH00133                          PORTMAN, ROB THE HONORA       [REP]   
34  S2HI00106                                  HIRONO, MAZIE K       [DEM]   
35  S2HI00106                                 HIRONO, MAZIE K.       [DEM]   
36  S2HI00106                                HIRONO, MAZIE MRS       [DEM]   
37  S6WY00068                                   BARRASSO, JOHN       [REP]   
38  S6WY00068                                 BARRASSO, JOHN A       [REP]   
39  P20000642                       CLINTON, WILLIAM JEFFERSON       [DEM]   
40  S2MN00126                                    COLEMAN, NORM       [REP]   
41  S2NC00083                                  DOLE, ELIZABETH       [REP]   
42  S2NC00083                                DOLE, ELIZABETH H       [REP]   
43  S2SD00068                                      THUNE, JOHN       [REP]   
44  S2SD00068                                    THUNE, JOHN R       [REP]   
45  S6OH00163                                   BROWN, SHERROD       [DEM]   
46  S8NC00239                                     HAGAN, KAY R       [DEM]   
47  S8VA00214                                    ALLEN, GEORGE       [REP]   
48  P00003392                          CLINTON, HILLARY RODHAM       [DEM]   
49  P00003392  CLINTON, HILLARY RODHAM / TIMOTHY MICHAEL KAINE       [DEM]   
50  P80003353               

### Cardinality of Candidates to Committees, through `cand_cmte` intersect table ###

Now see how many Candidates do not have a Committee association through the `cand_cmte` intersect table

In [56]:
%%sql result <<
select count(distinct cand_id) orphans_distinct,
       round(count(distinct cand_id)::numeric / :cand_distinct_ids * 100.0, 2) as pct_distinct_cands
  from cand
 where cand_id not in
       (select cand_id
          from cand_cmte)

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.
Returning data to local variable result


In [57]:
orphans_distinct = int(result.loc[0][0])
result

orphans_distinct pct_distinct_cands
0              7960              30.33

Let's see if the unassociated nature of these Candidate records has something to do with their "status"

In [58]:
%%sql
select cand_status,
       count(distinct cand_id) orphans_distinct,
       round(count(distinct cand_id)::numeric / :orphans_distinct * 100.0, 2) as pct_distinct_orphans
  from cand
 where cand_id not in
       (select cand_id
          from cand_cmte)
 group by 1
 order by 1

 * postgresql+psycopg2://crash@caladan/fecdb
5 rows affected.


cand_status  orphans_distinct pct_distinct_orphans
0           A                 1                 0.01
1           C                30                 0.38
2           F                 2                 0.03
3           N              7910                99.37
4           P                33                 0.41

Now let's look at all of the cardinalities of Candidate to Committee (minus orphans); we count the number of Candidates with various numbers of associations through `cand_cmte`

In [59]:
%%sql
with cand_cmte_card as (
    select cn.cand_id,
           count(distinct cm.cmte_id) as num_cmtes
      from cand cn
      left join cand_cmte cc on cc.cand_id = cn.cand_id
      left join cmte cm      on cm.cmte_id = cc.cmte_id
     group by 1
    having count(distinct cm.cmte_id) > 0
)
select num_cmtes as assoc_cmtes,
       count(*)  as cand_count,
       round(count(*)::numeric / (:cand_distinct_ids - :orphans_distinct) * 100.0, 2) as pct_non_orphans
    from cand_cmte_card
 group by 1
 order by 1 desc

 * postgresql+psycopg2://crash@caladan/fecdb
39 rows affected.


assoc_cmtes  cand_count pct_non_orphans
0            51           1            0.01
1            47           1            0.01
2            44           1            0.01
3            43           2            0.01
4            41           1            0.01
5            38           1            0.01
6            35           1            0.01
7            34           1            0.01
8            32           2            0.01
9            31           1            0.01
10           30           1            0.01
11           29           4            0.02
12           28           2            0.01
13           27           1            0.01
14           25           5            0.03
15           24           1            0.01
16           23           5            0.03
17           22           2            0.01
18           21           4            0.02
19           20           5            0.03
20           19           3            0.02
21           18           7            0.04
22           17           6            0.03
23           16          11            0.06
24           15           9            0.05
25           14          10            0.05
26           13           9            0.05
27           12          19            0.10
28           11          17            0.09
29           10          21            0.11
30            9          33            0.18
31            8          56            0.31
32            7          57            0.31
33            6          57            0.31
34            5         111            0.61
35            4         185            1.01
36            3         368            2.01
37            2        1460            7.99
38            1       15802           86.43

Let's see who the candidates are that have 20 or more associated Committees (though `cand_cmte`), including the office(s) they are running for

In [60]:
%%sql
with cand_cmte_card as (
    select cn.cand_id,
           count(distinct cm.cmte_id) as num_cmtes
      from cand cn
      join cand_cmte cc on cc.cand_id = cn.cand_id
      join cmte cm      on cm.cmte_id = cc.cmte_id
     group by 1
    having count(distinct cm.cmte_id) >= 20
)
select cn.cand_id,
       cn.cand_name,
       array_agg(distinct cn.cand_pty_affiliation) as party,
       ccc.num_cmtes,
       array_agg(distinct concat('''', cn.cand_election_yr, ' ', cn.cand_office_st, ' ', cn.cand_office, ''''))
           as "race (year state office)"
  from cand_cmte_card ccc
  join cand cn on cn.cand_id = ccc.cand_id
 group by 1, 2, 4
 order by 4 desc

 * postgresql+psycopg2://crash@caladan/fecdb
55 rows affected.


cand_id                        cand_name       party  num_cmtes  \
0   S6MO00305                MCCASKILL, CLAIRE       [DEM]         51   
1   P00003335                   BUSH, GEORGE W       [REP]         47   
2   S8NC00239                     HAGAN, KAY R       [DEM]         44   
3   S0NH00219                  SHAHEEN, JEANNE       [DEM]         43   
4   S6LA00227                 LANDRIEU, MARY L       [DEM]         43   
5   S6MN00267                   KLOBUCHAR, AMY       [DEM]         41   
6   S6MN00267                 KLOBUCHAR, AMY J  [DEM, DFL]         41   
7   S2WI00219                   BALDWIN, TAMMY       [DEM]         38   
8   P80003338                    OBAMA, BARACK       [DEM]         35   
9   P80003338  OBAMA, BARACK / JOSEPH R. BIDEN       [DEM]         35   
10  S8WA00194                  CANTWELL, MARIA       [DEM]         34   
11  S6MT00162                      TESTER, JON       [DEM]         32   
12  S6MT00162                   TESTER, R. JON       [DEM]         32   
13  S8MI00281                 STABENOW, DEBBIE       [DEM]         32   
14  P20000527                     NADER, RALPH  [IND, UNK]         31   
15  P60002458                  JEWELL, ROGER H       [REP]         30   
16  P60002458              JEWELL, ROGER HENRY  [NNE, REP]         30   
17  P80000235                    KERRY, JOHN F       [DEM]         29   
18  P80002801                   MCCAIN, JOHN S       [REP]         29   
19  P80002801                  MCCAIN, JOHN S.       [REP]         29   
20  S2ND00099                  HEITKAMP, HEIDI       [DEM]         29   
21  S4NC00089                    BURR, RICHARD       [REP]         29   
22  S4NC00089                  BURR, RICHARD M       [REP]         29   
23  S2CA00286                   BOXER, BARBARA       [DEM]         28   
24  S8CO00172                    UDALL, MARK E       [DEM]         28   
25  S8FL00166                     NELSON, BILL       [DEM]         27   
26  S0CA00199                FEINSTEIN, DIANNE       [DEM]         25   
27  S0NH00235                  AYOTTE, KELLY A       [REP]         25   
28  S0OH00133                     PORTMAN, ROB       [REP]         25   
29  S0OH00133          PORTMAN, ROB THE HONORA       [REP]         25   
30  S6OH00163                   BROWN, SHERROD       [DEM]         25   
31  S8AK00090                     BEGICH, MARK       [DEM]         25   
32  S8VA00214                    ALLEN, GEORGE       [REP]         24   
33  P80000912                         GORE, AL       [DEM]         23   
34  S0MO00183                       BLUNT, ROY       [REP]         23   
35  S2TX00106                     CORNYN, JOHN       [REP]         23   
36  S2TX00106                 CORNYN, JOHN SEN       [REP]         23   
37  S4PA00121           TOOMEY, PATRICK JOSEPH       [REP]         23   
38  S6PA00217               CASEY, ROBERT P JR       [DEM]         23   
39  S6PA00217             CASEY, ROBERT P. JR.       [DEM]         23   
40  P80000748                        PAUL, RON       [REP]         22   
41  S0CO00211                BENNET, MICHAEL F       [DEM]         22   
42  S0NY00410    GILLIBRAND, KIRSTEN ELIZABETH       [DEM]         21   
43  S2MN00126                    COLEMAN, NORM       [REP]         21   
44  S2SD00068                      THUNE, JOHN       [REP]         21   
45  S2SD00068                    THUNE, JOHN R       [REP]         21   
46  S8MT00119                    BURNS, CONRAD       [REP]         21   
47  H8OH01043                    CHABOT, STEVE       [REP]         20   
48  H8OH01043                 CHABOT, STEVEN J       [REP]         20   
49  H8OH01043            CHABOT, STEVEN JOSEPH       [REP]         20   
50  S2NV00209                 BERKLEY, SHELLEY       [DEM]         20   
51  S6OR00094             SMITH, GORDON HAROLD       [REP]         20   
52  S6WY00068                   BARRASSO, JOHN       [REP]         20   
53  S6WY00068                 BARRASSO, JOHN A       [REP]      

## Summary ##

### Findings ###

* `cmte.cand_id` foreign key
    * High integrity, more than 99.9% of keys successfully join to `cand`
    * In a small number of instances (less than 0.5%), the `cmte.cand_id` key changes for a Committee record (identified by `cmte.cmte_id`) across election cycles; visual examination shows that in almost all cases, it is between Candidate records with identical or very similar names (and thus, the same underlying person)
    * For Candidate records not referenced by a Committee through a direct join ("orphans"), the vast majority of them (over 99%) have a `cand_status` value of "N"
        * Of the "orphans" with `cand_status` of "C", the reverse join through `cand.cand_pcc` successfully identifies a valid Committee about 50-70% of the time (depending on how the orphan is identified), including a closely matching Candidate record (joined to the Committee through `cmte.cand_id`)
* `cand.cand_pcc` foreign key
    * High integrity (99.7% of keys successfully join to `cmte`) when `cand.cand_status` has values of "C" (statutory candidate), "P" (prior statutory candidate), or "F" (future statutory candidate)
    * Low integrity (59% of keys successfully join to `cmte`) when `cand.cand_status` has a value of "N" (not yet a statutory candidate)
    * For successful joins across all `cand.cand_status` values, there is a matching reciprocal join from `cmte` to `cand` through `cmte.cand_id` more than 99% of the time
        * Looking at `cand.cand_status` = "C", the non-reciprocal joins (only 0.5% of cases) went to other Candidate records with the **same name** 63% of the time, and Candidate records with **different names** (though almost always similar) 37% of the time
* `cand_cmte` intersect table
    * Integrity of the foreign keys is high, 100% valid for `cmte_id` and 99.97% valid for `cand_id`
    * 0.5% of the records have duplicate `cmte_id` and `cand_id` associations within an election cycle; further analysis shows that it is almost entirely due to differing values for `cand_cmte.cand_election_yr` (not sure how this field is used&mdash;note that many of the diffs contain a value of an odd-numbered year)
    * 3.2% of Committee records that join to Candidate through `cand_cmte` join to multiple Candidate records; of those, 81% join to 2, 3, or 4 Candidates (though we don't currently know how those Candidate records relate to one another&mdash;e.g. whether they represent the same, or disparate, real-world people)
    * In almost all cases of joins between Committee and Candidate through `cand_cmte`, the direct join through `cmte.cand_id` matches one of the intersect table records
        * In the case of one-to-one associations through `cand_cmte`, the match rate is 99.98%
    * For Candidate records not referenced by a Committee through the intersect table, the vast majority of them (over 99%) have a `cand_status` value of "N" (as with the direct association using `cmte.cand_id`)

### Recommendations ###

Based on the above exploration and findings, the following enhancements to the schema are recommended:

* Create `cand_cmte_mstr` table based on unique `cmte_id`/`cand_id` associations (across election cycles)
    * Still need to explore a little more whether specificity based on any other `cand_cmte` attributes is warranted (currently leaning against this idea)
* Create "Base Candidate" associations (through `cand_mstr.base_cand_mstr_id`, see `dc2` notebook) based on Candidate records that are closely-related through various relationships with common Committee records
